aptos-move/move-examples/nfts_as_accounts/sources/hero.move

```diff
@@ -0,0 +1,287 @@
+/// This demonstrates how to make a composable NFT on Aptos. We take a Hero that can have optional
+/// weapons, armor, and a shield. Each of these can optionally have a gem. These can be added or
+/// removed and seamlessly treated as Tokens or their richer type.
+///
+/// todo(@davidiw):
+/// * complete armor and shield
+/// * add accessors to everything
+/// * make it possible to get a reference to the various ref types so that the items can remain
+///   as is without requiring decomposing
+/// * better error handling around if an item is already equipped
+/// * add the ability to remove weapons and gems
+module nfts_as_accounts::composable_example {
+    use std::option::{Self, Option};
+    use std::signer;
+    use std::string::{Self, String};
+
+    use nfts_as_accounts::token;
+
+    const ENOT_A_HERO: u64 = 1;
+    const ENOT_A_WEAPON: u64 = 2;
+    const ENOT_A_GEM: u64 = 3;
+
+    struct OnChainConfig has key {
+        collection: String,
+        mutability_config: token::MutabilityConfig,
+        royalty: token::Royalty,
+    }
+
+    struct Hero has key {
+        armor: Option<ArmorRef>,
+        gender: String,
+        race: String,
+        shield: Option<ShieldRef>,
+        weapon: Option<WeaponRef>,
```

Comment:
```
// attempt 2
struct Character {
  // perhaps Alice anticipates the need for a Sword and a Shirt up front...
  sword: Option<Sword>,
  shirt: Option<Shirt>
  // ...but what happens when Daniel comes along later and wants to add Pants?
}
```

Reference: Sample code in the [Why we created SUI Move](https://medium.com/mysten-labs/why-we-created-sui-move-6a234656c36b)

Reply:
Just like the example shown above, there is no real constraint here. A dev could emulate the `TokenStore`, e.g., if a dev wanted to add `accessories: Option<TokenRef>`, they totally could. Most games, I've played, however, have a limited number of asset types so that they can constrain the development experience. Generally speaking, completely arbitrary types aren't easily programmable across the myriad domains used. For example, cool, the hero in a "Hero Quest" has pants, but the game devs don't even know what a pants object is. What do they do with it? It just wastes space on the hero.

Also Aptos has [any](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/any.move), but I'm avoiding it because I wanted something that was a bit more performant and inherently typesafe.

Reply:
Could you please demonstrate it more with the combination of Aptos upgradable contract ability? For example, adding a pant for the hero in the new version of the game.

Reply:
Well if you wanted to guarantee upgradeability of that sorts, then all Move implementations similarly suffer. You must either support a heterogeneous bag have functions that only support the types that you're willing to let it contain or go the route of migrating data types.

For the heterogeneous approach, one could theoretically just add new functions that specify what type of objects can be added. So I could add new functions like `fun wear_pants` and `fun remove_pants`. Those would then have the internal logic to convert from `PantsRef` to `TokenRef` and store the `TokenRef` into the bag.